### PR TITLE
Add time based rotating hash to payload

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/23055

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

`rotating_hash` should appear in the payload when redeeming unblinded tokens on desktop. This hash should be the same across multiple instances of the browser on the same OS and should rotate (change) every hour. `rotating_hash` should not be sent in the payload on Android or iOS. The rotating hash is built from the device id, creative instance id and a rotating timestamp. You can view the device id at brave://rewards-internals > Ads Diagnostics > Device Id. We should also confirm the device id is different on different OS instances including multiple instances on VMs.

### CODE REVIEW

**Comment by tmancey** [comment]
This is a POC and once/if approved from privacy team we will then fix the layering violations

File: components/brave_ads/browser/ads_service_impl.cc [hunk 1]
```diff
@@ -830,15 +835,18 @@ void AdsServiceImpl::OnResetAllState(const bool success) {
   VLOG(1) << "Successfully reset ads state";
 }
 
-void AdsServiceImpl::DetectUncertainFuture(const uint32_t number_of_start) {
+void AdsServiceImpl::DetectUncertainFuture(const uint32_t number_of_start,
+                                           const std::string& device_id) {
   auto callback = base::BindOnce(&AdsServiceImpl::OnDetectUncertainFuture,
-                                 AsWeakPtr(), number_of_start);
+                                 AsWeakPtr(), number_of_start, device_id);
```
Comment (aseren) [inline components/brave_ads/browser/ads_service_impl.cc:841]: We are passing the reference to a temporary object here. It is safer to pass std::string by value.

File: vendor/bat-native-ads/src/bat/ads/internal/account/user_data/rotating_hash_user_data.cc [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+constexpr char kRotatingHashKey[] = "rotating_hash";
+}  // namespace
+
+base::DictionaryValue GetRotatingHash(const std::string& creative_instance_id) {
+  base::DictionaryValue user_data;
+
+  const std::string& device_id = SysInfo().device_id;
+
+  const int64_t timestamp_rounded_to_nearest_hour =
+      static_cast<int64_t>(base::Time::Now().ToDoubleT()) /
```
Comment (aseren) [inline vendor/bat-native-ads/src/bat/ads/internal/account/user_data/rotating_hash_user_data.cc:27]: maybe we could do conversion after division?

File: vendor/bat-native-ads/src/bat/ads/internal/account/user_data/rotating_hash_user_data.cc [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+      base::Time::kSecondsPerHour;
+  const std::string timestamp =
+      base::NumberToString(timestamp_rounded_to_nearest_hour);
+
+  const std::string rotating_hash = base::NumberToString(
+      base::Hash(device_id + creative_instance_id + timestamp));
```
Comment (aseren) [inline vendor/bat-native-ads/src/bat/ads/internal/account/user_data/rotating_hash_user_data.cc:33]: base::StrCat could be used here

File: components/brave_ads/browser/ads_service_impl.cc (hunk 1 above)
Comment (aseren) [reply components/brave_ads/browser/ads_service_impl.cc]: fixed

File: vendor/bat-native-ads/src/bat/ads/internal/account/user_data/rotating_hash_user_data.cc (hunk 2 above)
Comment (aseren) [reply vendor/bat-native-ads/src/bat/ads/internal/account/user_data/rotating_hash_user_data.cc]: fixed

File: vendor/bat-native-ads/src/bat/ads/internal/account/user_data/rotating_hash_user_data.cc (hunk 3 above)
Comment (aseren) [reply vendor/bat-native-ads/src/bat/ads/internal/account/user_data/rotating_hash_user_data.cc]: fixed

**Review by aseren — DISMISSED** [review]

**Comment by fmarier** [comment]
@tmancey Please add this new field to https://github.com/brave/ads-serve/wiki/Data-Collection once the PR has been merged. Thanks.

File: browser/brave_ads/device_id/device_id_impl.h [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+#define BRAVE_BROWSER_BRAVE_ADS_DEVICE_ID_DEVICE_ID_IMPL_H_
+
+#include <stddef.h>
+
+#include "base/task/task_traits.h"
+#include "brave/components/brave_ads/browser/device_id.h"
+
+namespace brave_ads {
+
+class DeviceIdImpl : public DeviceId {
```
Comment (goodov) [inline browser/brave_ads/device_id/device_id_impl.h:16]: why not use `components/metrics/machine_id_provider.h`?

**Comment by tmancey** [comment]
@goodov as unfortunately that can be turned off from the command line args

**Comment by goodov** [comment]
> @goodov as unfortunately that can be turned off from the command line args

ok. my question was about `components/metrics/machine_id_provider.h`, but I deleted it because it just returns `hardware model name. (e.g. 'Macbook Pro 16,1', 'iPhone 9,3')` on non-windows machines.

**Comment by tmancey** [comment]
Ah thanks, did not notice it was deleted as on my phone

**Review by aseren — APPROVED** [review]

**Review by goodov — APPROVED** [review]
`DEPS` lgtm

**Pull request merged**